lib/core/src/Cardano/Wallet/Network.hs

```diff
@@ -378,26 +378,33 @@ follow'
     -- ^ Getter on the abstract 'block' type
     -> IO FollowExit
 follow' nl tr cps yield header =
-    bracket (initCursor nl cps) (destroyCursor nl) (sleep 0 False)
+    bracket
+        (initCursor nl cps)
+        (destroyCursor nl)
+        (handleExceptions . sleep 0 False)
   where
     innerTr = contramap MsgFollowLog tr
     delay0 :: Int
     delay0 = 500*1000 -- 500ms
 
+    handleExceptions :: IO FollowExit -> IO FollowExit
+    handleExceptions =
+        -- Node disconnections are seen as async exceptions from here. By
+        -- catching them, `follow` will try to establish a new connection
+        -- depending on the `FollowExceptionRecovery`.
+        handleSyncOrAsync (traceException *> const (pure FollowFailure))
```

Comment:
I really doubt that we need to handle asynchronous exceptions.
Just `UnliftIO.Exception.handleAny` (any synchronous exception) should be enough.

Reply:
We already _used to_ catch synchronous exceptions.

Note that chain-sync is run on a separate thread from `follow`, connected via [`link`](https://github.com/input-output-hk/cardano-wallet/blob/2738c71c44466d99804acb455a15ae8a5cb51cd7/lib/shelley/src/Cardano/Wallet/Shelley/Network.hs#L422), which seems to indeed cause async exceptions to be thrown:

```haskell
λ> f = async (threadDelay 1000000 >> throwIO (userError "hi")) >>= link >> threadDelay 3000000

λ> f `catchSyncOrAsync` (\(e::SomeException) -> putStrLn "Caught")
Caught

λ>  f `catchAny` (\_e -> putStrLn "Caught")
*** Exception: ExceptionInLinkedThread (ThreadId 988) user error (hi)
```

Reply:
OK, then we should either be catching the exceptions in the chain sync thread, or not using `link`. Trying to catch exceptions from other threads is a recipe for disaster.

Reply:
Ok 👍